pkg/reconcilermanager/controllers/rootsync_controller.go

```diff
@@ -779,52 +793,149 @@ func (r *RootSyncReconciler) validateRootSecret(ctx context.Context, rootSync *v
 	return validateSecretData(rootSync.Spec.Auth, secret)
 }
 
-func (r *RootSyncReconciler) configureClusterRoleBinding(ctx context.Context, reconcilerRef types.NamespacedName, roleConfig v1beta1.ClusterRoleOverrideSpec) error {
-	// in order to be backwards compatible with behavior before https://github.com/GoogleContainerTools/kpt-config-sync/pull/938
-	// we delete any ClusterRoleBinding that uses the old name.
-	// This ensures smooth migrations for users upgrading past that version boundary.
-	if err := r.deleteClusterRoleBinding(ctx, fmt.Sprintf("%s:%s", core.RootReconcilerPrefix, configsync.RootSyncName)); err != nil && !apierrors.IsNotFound(err) {
-		return errors.Wrap(err, "deleting old binding")
+func (r *RootSyncReconciler) listCurrentRoleRefs(ctx context.Context, rsRef types.NamespacedName) (map[v1beta1.RootSyncRoleRef]client.Object, error) {
+	currentRoleMap := make(map[v1beta1.RootSyncRoleRef]client.Object)
+	labelMap := r.managedObjectLabelMap(rsRef)
+	opts := &client.ListOptions{}
+	opts.LabelSelector = client.MatchingLabelsSelector{
+		Selector: labels.SelectorFromSet(labelMap),
+	}
+	rbList := rbacv1.RoleBindingList{}
+	if err := r.client.List(ctx, &rbList, opts); err != nil {
+		return nil, errors.Wrap(err, "listing RoleBindings")
+	}
+	for idx, rb := range rbList.Items {
+		roleRef := v1beta1.RootSyncRoleRef{
+			RoleRefBase: v1beta1.RoleRefBase{
+				Kind: rb.RoleRef.Kind,
+				Name: rb.RoleRef.Name,
+			},
+			Namespace: rb.Namespace,
+		}
+		currentRoleMap[roleRef] = &rbList.Items[idx]
 	}
-
-	if roleConfig.Disabled {
-		return errors.Wrap( // wrap returns nil if err is nil
-			r.deleteClusterRoleBinding(ctx, RootSyncPermissionsName(reconcilerRef.Name)),
-			"ensuring clusterrolebinding does not exist",
-		)
+	crbList := rbacv1.ClusterRoleBindingList{}
+	if err := r.client.List(ctx, &crbList, opts); err != nil {
+		return nil, errors.Wrap(err, "listing ClusterRoleBindings")
+	}
+	for idx, crb := range crbList.Items {
+		roleRef := v1beta1.RootSyncRoleRef{
+			RoleRefBase: v1beta1.RoleRefBase{
+				Kind: crb.RoleRef.Kind,
+				Name: crb.RoleRef.Name,
+			},
+		}
+		currentRoleMap[roleRef] = &crbList.Items[idx]
 	}
+	return currentRoleMap, nil
+}
 
-	if roleConfig.Name == "" {
-		roleConfig.Name = "cluster-admin"
+func (r *RootSyncReconciler) cleanRBACBindings(ctx context.Context, reconcilerRef, rsRef types.NamespacedName) error {
```

Comment:
cleanRBACBindings -> cleanBindings ? This cleans Cluster and Namespaced RoleBindings, right?

Reply:
Yes this cleans both ClusterRoleBindings and RoleBindings, so I figured cleanRBACBindings would be a precise yet generic name